### ISSUE TITLE
`elf.sym['read']` gives an incorrect address on Ubuntu 20.04.3 LTS with gcc 9.3.0

### ISSUE THREAD

**Issue by aesophor** [Gallopsled/pwntools#1983]
## Environment
| Name | Version |
| --- | --- |
| OS | **Ubuntu 20.04.3 LTS (focal)** |
| Pwntools | **4.6.0** |
| gcc | **9.3.0-17ubuntu1~20.04** |
| GNU ld | **2.34** |

I compiled the following program using `gcc 9.3.0` (the default one on Ubuntu 20.04.3 LTS).

```c
#include <unistd.h>

int main() {
    // no brute forcing
    sleep(3);
    // exploit me
    char buf[4];
    read(0, buf, 0x100);
}
```

Here's the problem:
* pwntools: `hex( ELF('./unexploitable').sym['read'] )` gives `0x401054`
* gdb: `read@plt` is at `0x401050`

![Screen Shot 2021-10-09 at 5 10 42 PM](https://user-images.githubusercontent.com/32998585/136652355-bedda912-2717-442d-a1e9-f80db5f876c0.png)

However, I also compiled this program on Arch Linux using `gcc 11.1.0` and the result is correct.
This erroneous result only happens on **Ubuntu 20.04.3 LTS** with **gcc 9.3.0**

---
The source, makefile and the compiled executable is provided in this zip file: [unexploitable.zip](https://github.com/Gallopsled/pwntools/files/7315310/unexploitable.zip)

**Comment by caprinux** [comment]
I tried this on multiple binaries, not sure if it's intended but it seems like pwntools look for the first instruction in the PLT, and returns that address. 

For binaries without the **endbr64** instruction at the start of the PLT, pwntools return the exact PLT address.

For binaries with the **endbr64** instruction at the start of the PLT, pwntools skip that instruction and return the address of the next instruction. 

Googling up endbr64, it seems to mostly do nothing more than a nop or may also serve as a minor security feature to detect control flow violations. Hence it doesn't seem to matter whether it includes that instruction in the PLT address or not, especially since we are already exploiting the program, a security feature will be unnecessary for us. Hence I feel that this should not be a bug. 

We can let the maintainers chip in and see what they have to say about it though :)

**Comment by zachriggle** [comment]
So there are some good, and not so good, explanations for this.

Basically, the PLT is a no-man's land, where there are no rules and north is actually south.  The GOT and PLT, when nicely formed by a friendly compiler, are in the same order.  On some architectures, they are even consistently sized!

This does not match the CTF environment :(

What Pwntools does is to use Unicorn to emulate the instructions in the entire PLT section, and find the FIRST instruction that can be emulated that results in the corresponding GOT entry being invoked.

To quote HHGTG:

> "This had made many people very angry and has been widely regarded as a bad move."

Unfortunately, there's not a better way to do it automatically, without spending a huge effort or pulling in other tools (IDA, Binary Ninja, Ghidra, etc) that do a better job.

All of that said, I'd definitely be in favor of having a special case for amd64 emulation that we do NOT start a PLT entry on an endbr64 instruction -- since that makes little sense anyway.  If anybody (@caprinux or @aesophor etc) are interested in creating a PR for this, it would be much appreciated.

TLDR: `endbr64` didn't exist when I wrote this code, and I'm sorry for the confusion ❤️
